File: Data Structures & Algorithms/house-robber/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    rob(nums) {
        const cache = new Map()
        const robFrom = (i) => {
            if (i >= nums.length) return 0
            if (cache.has(i)) return cache.get(i)

            const v = nums[i] + Math.max(robFrom(i + 2), robFrom(i + 3))
            cache.set(i, v)
            return v
        }

        return Math.max(robFrom(0), robFrom(1))
    }
}

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        l = 0
        r = len(nums) - 1
        while r - l > 1:
            m = (l + r) // 2
            if nums[l] <= nums[m]:
                print("->", l, m, r)
                l = m
            else:
                print("<-", l, m, r)
                r = m
        return min(nums[0], nums[r])